bin/dart2js_aggregate_tests.dart

```diff
@@ -0,0 +1,172 @@
+      mode = Mode.args;
+      break;
+    case 'build':
+      mode = Mode.build;
+      break;
+    default:
+      mode = Mode.test;
+      break;
+  }
+
+  buildDart2jsAggregateTestYaml(mode);
+  final testPaths = parseAggregateTestPaths(mode);
+  if (mode == Mode.args) {
+    printArgs(testPaths, debug: debug);
+  } else {
+    await buildOrRunTests(mode, testPaths, debug: debug);
+  }
+}
+
+/// Run a build with a filter for just the dart test config (which should be
+/// quick).
+///
+/// We need this file to be up-to-date so we can build the correct command with
+/// build filters for each of the intended test paths.
+void buildDart2jsAggregateTestYaml(Mode mode) {
+  var executable = 'pub';
+  var args = [
+    'run',
+    'build_runner',
+    'build',
+    '--build-filter=test/dart_test.dart2js_aggregate.yaml'
+  ];
+  logIf(mode != Mode.args, 'Building dart2js aggregate test config...');
+  logIf(mode != Mode.args, '$executable ${args.join(' ')}');
+  var result = Process.runSync(executable, args);
+  logIf(result.exitCode != 0 || mode != Mode.args,
+      '${result.stderr}\n${result.stdout}');
+  if (result.exitCode != 0) {
+    exit(result.exitCode);
+  }
+}
+
+/// Parse the generated test config to get the paths for each test that will be
+/// run, so that we can generate the correct --build-filter args.
+///
+/// Returns the list of paths for all aggregate tests, or exits early if they
+/// could not be parsed.
+List<String> parseAggregateTestPaths(Mode mode) {
+  logIf(mode != Mode.args, '\nReading dart2js aggregate test config...');
+  final configFile = File('test/dart_test.dart2js_aggregate.yaml');
+  if (!configFile.existsSync()) {
+    stdout
+        .writeln(r'''dart2js aggregation is not enabled. Update your build.yaml:
+
+# build.yaml
+targets:
+  $default:
+    builders:
+      test_html_builder:
+        options:
+          dart2js_aggregation: true''');
+    exit(1);
+  }
+
+  final config =
+      loadYaml(configFile.readAsStringSync(), sourceUrl: configFile.uri);
+  List<String> paths;
+  try {
+    paths = List<String>.from(config['presets']['dart2js-aggregate']['paths']);
+  } catch (e, stack) {
+    stdout
+      ..writeln('Failed to read test paths from "${configFile.uri}')
+      ..writeln(e)
+      ..writeln(stack);
+    exit(1);
+  }
+  logIf(mode != Mode.args, 'Found ${paths.length} aggregate tests to run.');
+  return paths;
+}
+
+/// Returns a list of args to be passed to a build_runner command that will only
+/// build/run [tests].
+///
+/// The --release flag will be included unless [debug] is true.
+List<String> buildRunnerBuildArgs(List<String> testPaths, {bool debug}) => [
+      if (debug != true) '--release',
+      for (final path in testPaths)
+        '--build-filter=${p.setExtension(path, '.**')}',
+    ];
+
+/// Depending on [mode], either builds or builds and runs the aggregate tests
+/// using [buildArgs] and [testArgs].
+Future<void> buildOrRunTests(Mode mode, List<String> testPaths,
```

Comment:
The doThisOrThat feels a little ugly, but it's a test runner, there's limited value to extra polishing.

Reply:
Ended up fixing this, as well